### ISSUE TITLE
[Question] Conditional model output when using SGD

### ISSUE THREAD

**Issue by fleskovar** [cornellius-gp/gpytorch#1751]
I am trying to use a multitask gaussian process for Bayesian optimization with a dataset that has about 10,000 records. In order to speedup training, I am trying to use mini-batch SGD to find the hyperparameters of the model´s kernel and mean (for this I randomly sample my training data, set the training data of the model, calculate the marginal log-likelihood and step the optimizer).

Before moving the model into `eval()`, I load the whole data to the model by using  `set_train_data()`. However, whenever I perform predictions, I observe that the mean of the model does not match the training examples. Moreover, I would expect the width of confidence interval for the function to decrease when approaching the training points.

My guess is that this is related to how `MultitaskGaussianLikelihood` works. Is this the case? Since I am trying to use SGD, I would like to make the accuracy of the predictions of the training data in the conditioned model independent of the results achieved during training. Is this possible?

**Comment by wjmaddox** [comment]
Can you post an example of this failing? My a priori suspicion would be that the task indicators are getting mixed up somehow in the optimization procedure.

This approach seems pretty related to a recent paper that figured out how to get unbiased estimates from an SGD based strategy [here](https://papers.nips.cc/paper/2020/file/1cb524b5a3f3f82be4a7d954063c07e2-Paper.pdf).

**Comment by fleskovar** [comment]
The code below uses the code of the "GPyTorch Regression Tutorial" to illustrates the point of the predictive mean and confidence interval not "converging" towards the training points:

```python
import math
import torch
import gpytorch
from matplotlib import pyplot as plt

# Training data is 100 points in [0,1] inclusive regularly spaced
train_x = torch.linspace(0, 1, 5)
# True function is sin(2*pi*x) with Gaussian noise
train_y = torch.sin(train_x * (2 * math.pi)) + torch.randn(train_x.size()) * math.sqrt(0.04)

# We will use the simplest form of GP model, exact inference
class ExactGPModel(gpytorch.models.ExactGP):
    def __init__(self, train_x, train_y, likelihood):
        super(ExactGPModel, self).__init__(train_x, train_y, likelihood)
        self.mean_module = gpytorch.means.ConstantMean()
        self.covar_module = gpytorch.kernels.ScaleKernel(gpytorch.kernels.MaternKernel())

    def forward(self, x):
        mean_x = self.mean_module(x)
        covar_x = self.covar_module(x)
        return gpytorch.distributions.MultivariateNormal(mean_x, covar_x)

# initialize likelihood and model
likelihood = gpytorch.likelihoods.GaussianLikelihood()
model = ExactGPModel(train_x, train_y, likelihood)


# Find optimal model hyperparameters
model.train()
likelihood.train()

# Use the adam optimizer
optimizer = torch.optim.Adam(model.parameters(), lr=0.1)  # Includes GaussianLikelihood parameters

# "Loss" for GPs - the marginal log likelihood
mll = gpytorch.mlls.ExactMarginalLogLikelihood(likelihood, model)

training_iter = 1000
for i in range(training_iter):
    # Zero gradients from previous iteration
    optimizer.zero_grad()
    # Output from model
    output = model(train_x)
    # Calc loss and backprop gradients
    loss = -mll(output, train_y)
    loss.backward()
    print('Iter %d/%d - Loss: %.3f   lengthscale: %.3f   noise: %.3f' % (
        i + 1, training_iter, loss.item(),
        model.covar_module.base_kernel.lengthscale.item(),
        model.likelihood.noise.item()
    ))
    optimizer.step()



# Get into evaluation (predictive posterior) mode
model.eval()
likelihood.eval()

# Test points are regularly spaced along [0,1]
# Make predictions by feeding model through likelihood
with torch.no_grad():
    test_x = torch.linspace(0, 1, 1000)
    observed_pred = likelihood(model(test_x))
    #observed_pred = model(test_x)

with torch.no_grad():
    # Initialize plot
    f, ax = plt.subplots(1, 1, figsize=(4, 3))

    # Get upper and lower confidence bounds
    lower, upper = observed_pred.confidence_region()
    # Plot training data as black stars
    ax.plot(train_x.numpy(), train_y.numpy(), 'k*')
    # Plot predictive means as blue line
    ax.plot(test_x.numpy(), observed_pred.mean.numpy(), 'b')
    # Shade between the lower and upper confidence bounds
    ax.fill_between(test_x.numpy(), lower.numpy(), upper.numpy(), alpha=0.5)
    ax.set_ylim([-3, 3])
    ax.legend(['Observed Data', 'Mean', 'Confidence'])
    plt.show()
```
Expected result:
![image](https://user-images.githubusercontent.com/6884660/133316448-e5040c90-bac9-4d85-a19f-e746fa37328f.png)

What I get most of the times:
![image](https://user-images.githubusercontent.com/6884660/133316539-025b30e1-9d47-4e5d-9be0-b0d096bc976a.png)
My understanding is that the model does not follow the points because the model was poorly trained, however since I am conditioning the model on the training data, I would expect to obtain something similar to the first picture even if I don't perform any training on the model and use the default parameters.

Below is a toy example of the multitask model. It is not using SGD (haven't had a chance yet to implement the proper data loader) but should illustrate my point about the GP not matching the training data:
```python
from typing import Union
from torch import Tensor
from gpytorch.kernels import MaternKernel, IndexKernel
from gpytorch.means import MultitaskMean, ConstantMean
from gpytorch.lazy import KroneckerProductLazyTensor
from gpytorch.likelihoods import MultitaskGaussianLikelihood, GaussianLikelihood
from gpytorch.models import ExactGP
from botorch.models.gpytorch import GPyTorchModel
from gpytorch.distributions import MultitaskMultivariateNormal, MultivariateNormal
from botorch.sampling import MCSampler, SobolQMCNormalSampler
import torch
from botorch.fit import fit_gpytorch_model
from botorch.optim.fit import fit_gpytorch_torch

import numpy as np
from gpytorch.mlls import ExactMarginalLogLikelihood
import matplotlib.pyplot as plt

class MultiOutputMultiTaskGP(ExactGP, GPyTorchModel):

    def __init__(
        self,
        input_labels: list[str],
        output_labels: list [str],
        train_X: Tensor,
        train_Y: Tensor,
        likelihood: MultitaskGaussianLikelihood = None,
        rank = None,
        lik_rank = None,
        input_loc: Tensor = None,
        input_scale: Tensor = None,
        output_loc: Tensor = None,
        output_scale: Tensor = None,
        sampler: MCSampler = None,
        device: str = "cpu"      
    ) -> None:

        self.input_labels = input_labels
        self.output_labels = output_labels

        num_tasks = train_Y.shape[-1]
        batch_shape, ard_num_dims = train_X.shape[:-2], train_X.shape[-1]

        self.input_loc = input_loc if input_loc is not None else torch.zeros(torch.Size([ard_num_dims-1,])).to(device)
        self.input_scale = input_scale if input_scale is not None else torch.ones(torch.Size([ard_num_dims-1,])).to(device)
        self.output_loc = output_loc if output_loc is not None else torch.zeros(torch.Size([num_tasks,])).to(device)
        self.output_scale = output_scale if output_scale is not None else torch.ones(torch.Size([num_tasks,])).to(device)
        
        self.distribution_cls = MultitaskMultivariateNormal if num_tasks > -1 else MultivariateNormal

        if lik_rank is None:
            lik_rank = rank
                       
        self._validate_tensor_args(X=train_X, Y=train_Y)

        if likelihood is None:
            if num_tasks > -1:
                likelihood = MultitaskGaussianLikelihood(
                    num_tasks=num_tasks, 
                    rank=lik_rank if lik_rank is not None else 0,
                )
            else:
                likelihood = GaussianLikelihood()

        super(MultiOutputMultiTaskGP, self).__init__(train_X, train_Y, likelihood)
        self.sampler = sampler if sampler is not None else SobolQMCNormalSampler(num_samples=500, seed=0, resample=False)
        self._rank = rank if rank is not None else num_tasks
        self._num_outputs = num_tasks

        self.mean_module = MultitaskMean(
            ConstantMean(), num_tasks=num_tasks
        )

        self.data_kernel = MaternKernel()
        self.task_kernel = IndexKernel(num_tasks=len(torch.unique(train_X[..., -1])))
        self.output_kernel = IndexKernel(num_tasks=num_tasks)
        
        self.to(device)
        
    def forward(self, x: Tensor) -> Union[MultitaskMultivariateNormal, MultivariateNormal]:
        mean_x = self.mean_module(x)
        task_term = self.task_kernel(x[..., -1].long())
        data_and_task_x = self.data_kernel(x[..., :-1]).mul(task_term)
        output_x = self.output_kernel.covar_matrix
        covar_x = KroneckerProductLazyTensor(data_and_task_x, output_x)
        return self.distribution_cls(mean_x, covar_x)

device = ["cuda", "cpu"][0]

x_task_1 = np.linspace(0, 1, 100)
y_task_1 = sum([
    -x_task_1+2,
    0.5* np.sin(10*x_task_1),
    0.05* np.sin(50*x_task_1),
    0.1* np.sin(50*(x_task_1+0.2))
])
x_task_1 = np.stack([x_task_1, np.ones_like(x_task_1)], axis=1)

x_task_0 = np.concatenate([np.linspace(0, 0.3, 30), np.linspace(0.6, 1, 40)], axis=0)
y_task_0 = sum([
    -x_task_0+2,
    0.5* np.sin(10*x_task_0),
    0.05* np.sin(50*x_task_0),
    0.1* np.sin(50*(x_task_0+0.2))
    
])
x_task_0_idx = np.stack([x_task_0, np.zeros_like(x_task_0)], axis=1)

x_train = np.concatenate([x_task_0_idx, x_task_1], axis=0)
y_train = np.expand_dims(np.concatenate([y_task_0, y_task_1], axis=0), axis=-1)

x_train = torch.tensor(x_train).to(device)
y_train = torch.tensor(y_train).to(device)

model = MultiOutputMultiTaskGP(["x"], ["y"], x_train, y_train, device=device)
marginal_log_likelihood = ExactMarginalLogLikelihood(model.likelihood, model)

model.to(x_train)
model.to(device)
marginal_log_likelihood.to(device)
model.train()

# Training loop (not using SGD). Maxiter set to 20 to get "bad" hyperparameters
fit_gpytorch_model(marginal_log_likelihood, optimizer=fit_gpytorch_torch, options={"maxiter": 20})

model.eval()
model.likelihood.eval()
with torch.no_grad():
    x_test = np.linspace(0, 1, 1000)
    x_test_idx = np.stack([x_test, np.zeros_like(x_test)], axis=1)
    test_point = torch.tensor(x_test_idx).to(device)
    
    post = model.posterior(test_point)
    lower, upper = post.mvn.confidence_region()
    std = post.mvn.stddev
    mean = post.mean

    plt.plot(x_test, mean.cpu().numpy().reshape(-1))
    
    plt.fill_between(x_test,
        mean.cpu().numpy().reshape(-1) - 3*std.cpu().numpy().reshape(-1),
        mean.cpu().numpy().reshape(-1) + 3*std.cpu().numpy().reshape(-1),
        alpha=0.5)

    # Calculate the true values of the underlying function
    y_task_0_full = sum([
        -x_test+2,
        0.5* np.sin(10*x_test),
        0.05* np.sin(50*x_test),
        0.1* np.sin(50*(x_test+0.2))    
    ])

    plt.plot(x_test, y_task_0_full, "r--") # True function
    plt.scatter(x_task_0, y_task_0)  # Training points shown to the model

    plt.show()
```

Result:
![image](https://user-images.githubusercontent.com/6884660/133331766-e3bac004-e28a-464b-b4b3-0aed8a02f2ae.png)

**Comment by wjmaddox** [comment]
Ugh, yes I think I see what's going on. In the first example, the optimizer is pretty severely under-fitting and attributing all of the data's variation to noise --- more data (which will mean a larger minibatch size ultimately) will help as I tried playing around with it and 10 data points was more stable (pretty consistently fitting).


In the second example, just running longer solves your problem -- 80-100 steps in torch give reasonable results? I've found MTGPs often generally need longer steps due to the increased number of parameters from the `IndexKernel`. 

The larger takeaway is that a) dataset size matters for GPs and b) GPs tend to under-fit at first before fitting the data.

Also, in the second example, is the reason that you're not using botorch's [KroneckerMultiTaskGP](https://botorch.org/api/models.html#botorch.models.multitask.KroneckerMultiTaskGP) the lack of support for multi-output (aka batched) multi-task GPs? Asking b/c that model is going to have considerably more efficient posterior sampling (shameless plug :) ).

**Comment by fleskovar** [comment]
Thank you so much for looking into this. I guess my confusion is about the role of training. If I understand correctly, training should find the optimum parameters (e.g. lengthscale) for the kernel and mean of the prior. However, even if I make a bad job at finding these parameters, I should be able to get the conditional distribution on my training points so that the resulting model "passes" through the observed data.

I also thought that the noise parameter of the likelihood might be causing the problem described above. I tried getting my predictive posterior by just calling `model(test_x)` instead of `likelihood(.)`  but this didn't seem to improve results.

(I was hoping to take a look at the KroneckerMultiTaskGP once I finish understanding this issue!)

**Comment by wjmaddox** [comment]
Yes, so what's likely happening is that there's a local "optimum" of the parameters which has [short lengthscale, large noise] for the five data point case (an "underfit" GP). Hence why you're severely under-fitting. In this case, conditioning on the observed data (your posterior) will do not that much and the GP's posterior will tend to be like the prior.

Here, I profiled out the 5 data regime to visualize, note the large trough of low MLL for a noise around 0.5:

![image](https://user-images.githubusercontent.com/13008542/133439983-e149d934-e5e4-4683-961e-9b24e4548ed0.png)
I also truncated values at 3 b/c they skyrocket past there.

To prevent this from occuring, beyond what I was suggesting previously, you could also place a prior / constraint on the likelihood (which changes the optimization landscape) to prevent it from getting too large before optimizing.

**Comment by fleskovar** [comment]
Thank you so much for the explanation, it really helped me understand how things work under the hood. I have just run some tests with priors and constraints on the noise parameter and I am now able to get the results I was expecting.

Additionally, thank you for the link to the paper about SGD and the recommendation on the KroneckerMultiTaskGP (I might come back with additional questions about this later!).
